src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -942,3 +944,213 @@ fixCopyCommand(Archive *fout, PQExpBuffer copyBuf, TableInfo *tbinfo, bool isFro
 	}
 	destroyPQExpBuffer(q);
 }
+
+/*
+ * fixCursorForBbfSqlvariantTableData:
+ * Prepare custom cursor for all Babelfish tables with atleast one sql_variant
+ * datatype column to correctly dump sql_variant data.
+ */
+void
+fixCursorForBbfSqlvariantTableData(Archive *fout, TableInfo *tbinfo, PQExpBuffer buf, int *nfields, bool *is_sqlvariant_column, bool *has_sqlvariant_column)
+{
+	int nfields_new = 0;
+	*(has_sqlvariant_column) = hasSqlvariantColumn(tbinfo);
+	
+	if(!has_sqlvariant_column || !isBabelfishDatabase(fout))
+		return;
+
+	resetPQExpBuffer(buf);
+	appendPQExpBufferStr(buf, "DECLARE _pg_dump_cursor CURSOR FOR SELECT ");
+
+	for (int i = 0; i < tbinfo->numatts; i++)
+	{
+		if (tbinfo->attisdropped[i])
+			continue;
+		if (tbinfo->attgenerated[i] && fout->dopt->column_inserts)
+			continue;
+
+		/* Skip TSQL ROWVERSION/TIMESTAMP column, it should be re-generated during restore. */
+		if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"rowversion\"" : "sys.rowversion") == 0 ||
+			pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"timestamp\"" : "sys.timestamp") == 0)
+			continue;
+
+		if (nfields_new > 0)
+			appendPQExpBufferStr(buf, ", ");
+		if (tbinfo->attgenerated[i])
+			appendPQExpBufferStr(buf, "NULL");
+		else
+		{
+			appendPQExpBufferStr(buf, fmtId(tbinfo->attnames[i]));
+			/*
+			 * To find the basetype and bytelength of string data types we
+			 * invoke sys.sql_variant_property and sys.datalength function on
+			 * the sqlvariant column. These extra columns are carefully handled
+			 * so that they do not interfere with expected dump behaviour.
+			*/
+			if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"sql_variant\"" : "sys.sql_variant") == 0)
+			{
+					appendPQExpBuffer(buf, ", sys.SQL_VARIANT_PROPERTY(%s, 'BaseType')", tbinfo->attnames[i]);
+					appendPQExpBuffer(buf, ", sys.datalength(%s)", fmtId(tbinfo->attnames[i]));
+					is_sqlvariant_column[nfields_new] = true;
+					nfields_new = nfields_new + 2;
+			}
+		}
+		nfields_new++;
+	}
+	*(nfields) = nfields_new;
+	return;
+}
+
+/*
+ * fixBbfSqlvariantData:
+ * Modify the values in INSERT query that dump sql_variant column table data so
+ * that metadata is preserved during restore. If default value is not used or 
+ * the value is not NULL, we add a cast expression to typecast the value to its
```

Comment:
Some corrections are needed:
1. Looks like this function just works on a single value at a time:
`s/Modify the values in INSERT query that dump sql_variant column table data so that metadata is preserved during restore/The function Modifies the value in INSERT command for sql_variant column so that metadata is preserved during restore/g`
`s/unless the column is generated or the value is NULL/If default value is not used or the value is not NULL/g`